libsolidity/formal/CHC.cpp

```diff
@@ -1512,16 +1543,27 @@ map<unsigned, vector<unsigned>> CHC::summaryCalls(CHCSolverInterface::CexGraph c
 {
 	map<unsigned, vector<unsigned>> calls;
 
-	solidity::util::BreadthFirstSearch<pair<unsigned, unsigned>>{{{_root, _root}}}.run([&](auto info, auto&& _addChild) {
-		auto [node, root] = info;
-		if (Predicate::predicate(_graph.nodes.at(node).name)->isSummary())
+	auto compare = [&](unsigned _a, unsigned _b) {
+		return _graph.nodes.at(_a).name > _graph.nodes.at(_b).name;
+	};
+
+	queue<pair<unsigned, unsigned>> q;
+	q.push({_root, _root});
+	while (!q.empty())
+	{
+		auto [node, root] = q.front();
+		q.pop();
+
+		Predicate const* nodePred = Predicate::predicate(_graph.nodes.at(node).name);
+		if (nodePred->isSummary() && (_root == root || nodePred->isInternalCall() || nodePred->isExternalCall()))
 		{
 			calls[root].push_back(node);
 			root = node;
 		}
-		for (auto v: _graph.edges.at(node))
-			_addChild({v, root});
-	});
+		auto const& edges = _graph.edges.at(node);
+		for (unsigned v: set<unsigned, decltype(compare)>(begin(edges), end(edges), compare))
+			q.push({v, root});
+	}
```

Comment:
So, what was the result of the discussion? :)
It is not possible to use one of the search utilities?

Reply:
Yea, here actually not. We *do* need a BFS, but we also need to visit certain nodes multiple times, which happens sometimes with loops.